design/20230726-cel-policy.md

```diff
@@ -0,0 +1,418 @@
+and I think it's one of the most common user stories the proposed feature will solve.
+
+#### SPIFFE ID identifying namespace
+
+As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing
+self-provisioning of [X.509 SVID certificates](https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md).
+The [SPIFFE ID](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md) (X.509 URI SAN) should identify
+the namespace (and service account), and must start with `spiffe://<trust-domain>/ns/<namespace>/sa/`.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go into as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes/PKI ecosystem.
+
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The proposal is to add sibling properties to `values` and `value` properties inside the `allowed` block. The new
+fields should contain CEL validations used to validate CSR attributes with similar
+[behavior](https://cert-manager.io/docs/projects/approver-policy/#allowed) as approver-policy has
+today, but with the following extension: If validations are present for the given CSR attribute, and **ALL**
+validations passes, the attribute will be allowed. This also holds if no values are listed.
+
+The validations themselves should be expressed using the same API as Kubernetes `x-kubernetes-validations`, having
+a mandatory `rule` field for the CEL expression and an optional `message` field for the validation error message.
+
+To use validations in CertificateRequestPolicy the API may look like this (example):
+
+```yaml
+spec:
+  allowed:
+    dnsNames:
+      validations:
+        - message: only dnsNames in the local namespace service domain is allowed.
+          rule: self.endsWith(cr.namespace + '.svc') || self.endsWith(cr.namespace + '.svc.cluster.local')
+```
+
+The CEL context variable `self` is inspired by
+[CEL expressions in Kubernetes](https://kubernetes.io/docs/reference/using-api/cel/),
+and should appear familiar to our users. 
+
+It should be allowed to mix `values` and `validations`, so the following policy should also be acceptable:
+
+```yaml
+spec:
+  allowed:
+    dnsNames:
+      validations:
+        - message: dnsName can be maximum 64 characters.
+          rule: self.size() =< 64
```

Comment:
How could we limit the number of dnsNames? Could we eg. have a "selfs" parameter?

Reply:
This is an interesting and important comment! Strictly speaking, many CSR attributes are multi-valued - so it could make sense to make `self` refer to the list of attribute values in the CSR - which will allow for the validation you are requesting here. But I suspect this will make the feature less user-friendly. The validation example this review comment refers to would then have to be expressed as `self.all(dnsName, dnsName.size() =< 64)` - which is not too bad IMO.

But this could easily lead to user errors like `self.size() =< 64` would now limit the number of dnsNames in the CSR to 63. 😉 Which is probably not what the user had in mind.

Reply:
IMO the most user-friendly is to keep the validations under `allowed`  as per-item validations. This is how the `value`/`values` fields in the current API work. More complex CEL validations, typically cross-cutting concerns could be introduced later outside of the `allowed` API structure. I am updating the proposal with this clarification now.